Oracle Database Support

pruivo opened the pull request:
Closes #70

- [x] Test Container
- [x] GH action

pruivo commented:
Unable to test the oracle container due to missing image `quay.io/gingersnap/cache-manager-oracle`

ryanemerson commented on src/main/java/io/gingersnapproject/metrics/micrometer/MicrometerMetrics.java:
```diff
@@ -62,6 +62,8 @@ void registerMetrics(@Observes Events.ConnectorStartedEvent ev) {
          case MYSQL -> registerMysqlConnectorMetrics(rule);
          case SQLSERVER -> registerGenericConnector(io.debezium.connector.sqlserver.Module.name(), rule);
          case POSTGRESQL -> registerGenericConnector(io.debezium.connector.postgresql.Module.name(), rule);
+         // TODO -> oracle metrics available in OracleStreamingChangeEventSourceMetricsMXBean
```
This PR or a follow up?

pruivo replied:
Follow-up, tracked by https://github.com/gingersnap-project/db-syncer/issues/71

ryanemerson commented on src/main/resources/oracle/populate.sql:
```diff
@@ -0,0 +1,37 @@
+-- All these insert/update/delete will be captured.
```
This isn't an issue with your PR specifically, but we should move the `populate.sql` files to `src/test`.

pruivo replied:
Follow up, tracked by https://github.com/gingersnap-project/db-syncer/issues/71

pruivo commented on deploy/oracle/oracle-compose.yaml:
```diff
@@ -0,0 +1,12 @@
+
+services:
+  oracledb:
+    image: gvenzl/oracle-xe:21-slim-faststart
```
To be honest, I didn't look for an official image. I used the image used by `OracleContainer` in `testcontainer`.
Do you prefer the official image?

ryanemerson replied:
I think it's probably the safer bet long term. I have know idea who/what `gvenzl` is and whether they're committed to maintaining the image long term.

That being said, if you can't get the official image to work with the `OracleContainer` or with the generic equivalent easily, then just proceed with what you have for now.

pruivo replied:
@ryanemerson my only concern is the official image is bigger:
```
gvenzl/oracle-xe                                     21-slim-faststart   91cd9b65ce8f   3 months ago    4.94GB
container-registry.oracle.com/database/express       21.3.0-xe           c273dde6b184   4 months ago    11.2GB
```
Configuration is slightly different (different env vars and directory locations), but not a big deal.
Should I continue with the change?

ryanemerson replied:
~ 5GB is massive,  11GB is insane!!!!!

Maybe we should stick with the unofficial one after all :thinking:

pruivo replied:
if you want a smaller image, there is the `21-slim`
```
gvenzl/oracle-xe                                     21-slim             855f24415eda   3 months ago    2.07GB
```
The difference between `21-slim` and `21-slim-faststart` seems minimal on my laptop. The  `21-slim` needs to uncompress some files:
```
oracle-oracledb-1  | CONTAINER: starting up...
oracle-oracledb-1  | CONTAINER: first database startup, initializing...
oracle-oracledb-1  | CONTAINER: uncompressing database data files, please wait...
oracle-oracledb-1  | CONTAINER: done uncompressing database data files, duration: 24 seconds.
oracle-oracledb-1  | CONTAINER: starting up Oracle Database...
```

ryanemerson replied:
I'm open to different opinions, but personally I want the smallest image possible. My 4G connection is not going to be happy otherwise :sweat_smile:

pruivo commented on deploy/oracle/setup.sh:
```diff
@@ -0,0 +1,79 @@
+  CREATE TABLE car_model(id INT NOT NULL PRIMARY KEY, model VARCHAR(255), brand VARCHAR(255));
+  ALTER TABLE car_model ADD SUPPLEMENTAL LOG DATA (ALL) COLUMNS;
+  exit;
+EOF
```
sure

pruivo commented:
Updated
* Created SQL files for Oracle setup
* Updated README
* Moved `populate.sql` to `src/test/resources`

Missing:
* Use official oracle image

ryanemerson commented:
@pruivo You also need to update the various GH actions to include the Oracle build.

pruivo commented:
> @pruivo You also need to update the various GH actions to include the Oracle build.

@ryanemerson Can you specify what needs to be updated? I'm unable to find anything.

ryanemerson commented:
> @ryanemerson Can you specify what needs to be updated? I'm unable to find anything.

Nope as I had commented on the wrong PR, this was meant for the cache-manager one :smile:

pruivo commented:
Updated:
* Enabled tests with Oracle

jabolina reviewed (approved):
Looks good.

pruivo commented on src/test/java/io/gingersnapproject/testcontainers/database/Oracle.java:
```diff
@@ -0,0 +1,56 @@
+import io.gingersnapproject.testcontainers.DatabaseProvider;
+
+/**
+ * TODO! Not tested!
```
Added javadoc

jabolina commented:
Merged, thanks @pruivo
